### PR TITLE
Forge 1.19.3 Updates & CreativeModeTabs

### PR DESCRIPTION
**Pull request by ApexModder**
This PR brings the needed changes to run on the recent Forge 1.19.3 versions & adds a new CreativeModeTab registration system to match Forge's CreativeModeTab events.

Updated data-gens to all take in `PackOutput` rather than `DataGenerator` to match Forge & Vanilla

Added methods to create, register & modify CreativeModeTabs
- `AbstractRegistrate#buildCreativeModeTab(String, [@Nullable List<Object>], [@Nullable List<Object>], Consumer<CreativeModeTab.Builder>, [@Nullable String])`
	- Used to build & register a new CreativeModeTab
	- Marks the first registered CreativeModeTab as the 'defaultTab', this is passed across to any 
'ItemBuilder' objects to pre-set the items display tab
	- Returns lazy-supplier to the registered instance (filled in later during registration)
	- Optional parameter to auto register translatable components for the tabs display name
- `AbstractRegistrate#creativeModeTab(String, [@Nullable List<Object>], [@Nullable List<Object>], Consumer<CreativeModeTab.Builder>, [@Nullable String])`
	- The same as `#buildCreativeModeTab` but returns the Registrate instance to allow method chaining
- `AbstractRegistrate#creativeModeTab(Supplier<? extends CreativeModeTab>)`
	- Used to set the default CreativeModeTab instance
- `AbstractRegistrate#modifyCreativeModeTab(Supplier<? extends CreativeModeTab>, Consumer<CreativeModeTabModifier>)`
	- Used to register callbacks which are used to modify what items are displayed on the given tab
- `ItemBuilder#tab(NonNullSupplier<? extends CreativeModeTab>, [Consumer<AbstractRegistrate.CreativeModeTabModifier>])`
	- Used to display the item built from this builder on the given tab
	- Optional consumer parameter to modify the ItemStack instance
- `ItemBuilder#removeTab(NonNullSupplier<? extends CreativeModeTab>)`
	- Used to remove the item built from this builder from the given tab

### CODE REVIEW

File: src/main/java/com/tterrag/registrate/AbstractRegistrate.java [hunk 1]
```diff
@@ -123,6 +124,79 @@ void addRegisterCallback(NonNullConsumer<? super T> callback) {
...
+
+        @Override
+        public CreativeModeTab get() {
+            return Objects.requireNonNull(creativeModeTab, () -> "Attempt to obtain CreativeModeTab(%s) instance before it was registered!".formatted(registryName));
+        }
+
+        @SuppressWarnings("unchecked")
+        private static List<Object> defaultAfterEntries() {
+            try {
+                var field = CreativeModeTabEvent.class.getDeclaredField("DEFAULT_AFTER_ENTRIES");
```
Comment (tterrag1098) [inline src/main/java/com/tterrag/registrate/AbstractRegistrate.java]: It is simpler to use `ObfuscationReflectionHelper` here, even if the target is not obfuscated.

File: src/main/java/com/tterrag/registrate/AbstractRegistrate.java [hunk 2]
```diff
@@ -123,6 +124,79 @@ void addRegisterCallback(NonNullConsumer<? super T> callback) {
         }
     }
 
+    public static final class CreativeModeTabModifier implements CreativeModeTab.Output {
```
Comment (tterrag1098) [inline src/main/java/com/tterrag/registrate/AbstractRegistrate.java]: I think there is enough here to extract this to its own class, and have the Registration class as package-private in the same file

File: src/main/java/com/tterrag/registrate/AbstractRegistrate.java [hunk 3]
```diff
@@ -582,38 +677,60 @@ public S object(String name) {
...
-     *
-     * @param tab
-     *            The tab to use for future items
-     * @return this {@link AbstractRegistrate}
-     */
-    @ApiStatus.ScheduledForRemoval(inVersion = "1.20")
-    @Deprecated(forRemoval = true, since = "1.19.3")
-    public S creativeModeTab(NonNullSupplier<? extends CreativeModeTab> tab) {
-        this.currentTab = Suppliers.memoize(tab::get);
+    public Supplier<CreativeModeTab> buildCreativeModeTab(String registryName, @Nullable List<Object> beforeEntries, @Nullable List<Object> afterEntries, Consumer<CreativeModeTab.Builder> configurator, @Nullable String englishTranslation) {
```
Comment (tterrag1098) [inline src/main/java/com/tterrag/registrate/AbstractRegistrate.java:658]: It's going to be a lot of boilerplate unfortunately, but these methods need some javadoc. Especially explaining the List<object> params. I'm reading the vanilla code that uses them and I'm still not 100% sure how they work...

File: src/main/java/com/tterrag/registrate/builders/ItemBuilder.java [hunk 4]
```diff
@@ -108,12 +109,29 @@ public static <T extends Item, P> ItemBuilder<T, P> create(AbstractRegistrate<?>
...
     private NonNullSupplier<Supplier<ItemColor>> colorHandler;
-    @Nullable private NonNullSupplier<? extends CreativeModeTab> tab = null;
-    private boolean registeredTabEvent = false;
+    private Map<NonNullSupplier<? extends CreativeModeTab>, Consumer<AbstractRegistrate.CreativeModeTabModifier>> creativeModeTabs = Maps.newHashMap();
 
     protected ItemBuilder(AbstractRegistrate<?> owner, P parent, String name, BuilderCallback callback, NonNullFunction<Item.Properties, T> factory) {
         super(owner, parent, name, callback, ForgeRegistries.Keys.ITEMS);
         this.factory = factory;
+
+        // delegate creative mode tab modification callback registration
```
Comment (tterrag1098) [inline src/main/java/com/tterrag/registrate/builders/ItemBuilder.java]: This is a nice explanation, but these important notes about how the tab methods work should be in javadoc on those methods.

**Review by tterrag1098 — CHANGES_REQUESTED** [review]

File: src/main/java/com/tterrag/registrate/AbstractRegistrate.java (hunk 2 above)
Comment (ApexModder) [reply src/main/java/com/tterrag/registrate/AbstractRegistrate.java]: I have split the modifier class off onto its own class, but unsure as to how would make the registration class package-private in the modifier class, as AbstractRegistrate needs visibility to it for the backend list field to keep track of registrations (`AbstractRegistrate#creativeModeTabsRegistrars`)

**Review by tterrag1098 — APPROVED** [review]

**Pull request closed without merging**